server/conn.go

```diff
@@ -284,36 +308,66 @@ func parseAttrs(data []byte) (map[string]string, error) {
 	return attrs, nil
 }
 
-func (cc *clientConn) readHandshakeResponse() error {
+func (cc *clientConn) readSSLRequestAndHandshakeResponse() error {
+	// Read a packet. It may be a SSLRequest or HandshakeResponse.
 	data, err := cc.readPacket()
 	if err != nil {
 		return errors.Trace(err)
 	}
 
-	var p handshakeResponse41
-	if err = handshakeResponseFromData(&p, data); err != nil {
+	var resp handshakeResponse41
+
+	pos, err := parseHandshakeResponseHeader(&resp, data)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// The packet is a SSLRequest, let's switch to TLS.
+	if (resp.Capability&mysql.ClientSSL > 0) && cc.server.cfg.SSLEnabled {
+		if err = cc.UpgradeToTLS(cc.server.tlsConfig); err != nil {
+			return errors.Trace(err)
+		}
+		// Read the following HandshakeResponse packet.
+		data, err = cc.readPacket()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		pos, err = parseHandshakeResponseHeader(&resp, data)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	// Read the remaining part of the packet.
+	if err = parseHandshakeResponseBody(&resp, data, pos); err != nil {
 		return errors.Trace(err)
 	}
-	cc.capability = p.Capability & defaultCapability
-	cc.user = p.User
-	cc.dbname = p.DBName
-	cc.collation = p.Collation
-	cc.attrs = p.Attrs
 
-	// Open session and do auth
-	cc.ctx, err = cc.server.driver.OpenCtx(uint64(cc.connectionID), cc.capability, uint8(cc.collation), cc.dbname)
+	cc.capability = resp.Capability & defaultCapability
```

Comment:
if server start with ssl but defaultCapability doesn't contain that bit, is it wrong here?